enhancements/agent-installer/automated-workflow-for-agent-based-installer.md

```diff
@@ -0,0 +1,616 @@
+- We want to install clusters with [single-node, compact, and
+  high-availability
+  topologies](/enhancements/single-node/cluster-high-availability-mode-api.md).
+- We want to support install configurations with platform `none`,
+  `baremetal`, or `vsphere`.
+- We want to avoid inventing completely new ways to describe OpenShift
+  clusters. Therefore we will reuse existing inputs to the installer
+  and to tools such as the zero-touch provisioning (ZTP) system in
+  multi-cluster engine (MCE).
+- On-premises deployments do not always have DHCP or other remote
+  services available to configure host networking. They also
+  frequently use more complex network configurations such as bonded
+  NICs and tagged VLANs. We therefore need to support configuration of
+  per-host network settings while installing.
+- On-premises deployments, especially on bare metal, may need to take
+  extra steps to fully utilize storage inside each host or visible to
+  a host from a SAN. We therefore need to support configuration of
+  per-host storage, including fibre channel settings such as
+  multipath.
+- We want to retain the OpenShift installer's ability to customize a
+  deployment with arbitrary manifests.
+
+### Non-Goals
+
+- A user-driven workflow will be described in a separate enhancement.
+- A user interface for collecting complex inputs such as per-host
+  network settings before running the installation automatically is
+  left to a future enhancement.
+- Changes to the release image mirroring process are out of scope for
+  this enhancement.
+- Eliminating the requirement for a separate image registry to host
+  the mirrored OpenShift image is left to a future enhancement.
+- Describing a mechanism to add and remove nodes from the cluster after
+  it is installed is left to a future enhancement.
+- Generating image formats other than ISO is left to a future
+  enhancement.
+- Booting hosts via PXE is left to a future enhancement.
+- Creating minimal ISO images with a separate rootfs is left to a
+  future enhancement.
+- We are not currently considering replacing the IPI installer or
+  assisted installer SaaS.
+- Describing an end-to-end solution for users without their own
+  orchestration tool is left to a future enhancement.
+- Collecting telemetry data for deployments is left to a
+  future enhancement.
+- Details of securing the communication between agents and services to
+  ensure only desired hosts join the cluster are left to a future
+  enhancement.
+- Details of a workflow for deploying in settings with DHCP assignment
+  of IP addresses is left to a future enhancement.
+- It is not currently a requirement that the downloadable form of the
+  installer support installing multiple versions of OpenShift. We will
+  address the underlying requirements behind that request in other
+  ways, which will be discussed elsewhere.
+
+## Proposal
+
+We propose to extend the `openshift-install` binary with a new
```

Comment:
Some discussion has been happening in a separate FAQ document. An important point came up about validations for optional operators that I think is relevant here. Paraphrasing, it was expressed that pre-installation validations that are specific to operators that handle storage, virt, etc. are more acceptable for the agent-based installation that is available on console.redhat.com, but would be less desirable for an installer that is shipped inside an openshift release due to coupling of requirements across products.

I think that's an important issue to address, especially if deciding to deliver this new install capability through the `openshift-install` binary is going to limit its value to users. Here was my comment:

> The main reason that a small number of operators are selectable on console.redhat.com is for the pre-install validations. If the hardware specs are not sufficient to run ODF, virt, etc, we want to tell the user upfront rather than let them go through a cluster installation, try to install and use the operator (disconnected operator installation is no fun...), and only then fail after all that time and work. The value of ensuring user success upfront was determined to be worth the price of maintaining a small list of validations for a few specific operators. IOW a small bit of coupling improves the UX significantly.

> Since a primary use case here is to enable a user who wants to establish a local point of management within their disconnected environment, validating their complete intent upfront seems to make sense. If putting this tool "inside OCP" precludes that sort of thinking, maybe that's not the best delivery model for the use case.

Were other delivery models considered and/or ruled out? What about shipping this as part of multi-cluster engine, where it could provide a complete story for creating a first local point of management? Or maybe we can still provide that complete story even while shipping as part of the OCP payload?

Reply:
That was an early decision that we captured in an internal document but didn't repeat here. It basically came down to wanting to integrate the various tools, rather than create another standalone thing. I'll share that doc with you.